serde_arrow/src/internal/serialization/flattened_union_builder.rs

```diff
@@ -0,0 +1,260 @@
+                .collect(),
+            row_count: self.row_count,
+        })
+    }
+
+    pub fn is_nullable(&self) -> bool {
+        false
+    }
+
+    pub fn into_array(self) -> Result<Array> {
+        let mut fields = BTreeMap::new();
+
+        for (builder, meta) in self.fields.into_iter() {
+            let ArrayBuilder::Struct(builder) = builder else {
+                fail!("enum variant not built as a struct") // TODO: better failure message
+            };
+
+            for (sub_builder, mut sub_meta) in builder.fields.into_iter() {
+                // TODO: this mirrors the field name structure in the tracer but represents
+                // implementation details crossing boundaries. Is there another way?
+                // Name change is currently needed for struct field lookup to work correctly.
+
+                sub_meta.name = format!("{}::{}", meta.name, sub_meta.name);
+                fields.insert(
+                    sub_meta.name.to_owned(),
+                    (sub_builder.into_array()?, sub_meta),
+                );
+            }
+        }
+
+        Ok(Array::Struct(StructArray {
+            len: self.row_count,
+            fields: fields.into_values().collect(),
+            // TODO: is this ok to hardcode?
+            // assuming so because when testing manually,
+            // validity of struct with nullable fields was None
+            validity: None,
+        }))
+    }
+}
+
+impl FlattenedUnionBuilder {
+    pub fn serialize_variant(&mut self, variant_index: u32) -> Result<&mut ArrayBuilder> {
+        let variant_index = variant_index as usize;
+
+        // don't serialize any variant not selected
+        for (idx, (builder, _meta)) in self.fields.iter_mut().enumerate() {
+            if idx != variant_index {
+                builder.serialize_none()?;
+            }
+        }
+
+        let Some((variant_builder, _variant_meta)) = self.fields.get_mut(variant_index) else {
+            fail!("Could not find variant {variant_index} in Union");
+        };
+
+        self.row_count += 1;
+
+        Ok(variant_builder)
+    }
+}
+
+impl Context for FlattenedUnionBuilder {
+    fn annotate(&self, annotations: &mut BTreeMap<String, String>) {
+        set_default(annotations, "field", &self.path);
+        set_default(annotations, "data_type", "Struct(..)");
+    }
+}
+
+impl SimpleSerializer for FlattenedUnionBuilder {
+    fn serialize_struct_variant_start<'this>(
+        &'this mut self,
+        _: &'static str,
+        variant_index: u32,
+        variant: &'static str,
+        len: usize,
+    ) -> Result<&'this mut ArrayBuilder> {
+        let mut ctx = BTreeMap::new();
+        self.annotate(&mut ctx);
+
+        try_(|| {
+            let variant_builder = self.serialize_variant(variant_index)?;
+            variant_builder.serialize_struct_start(variant, len)?;
+            Ok(variant_builder)
+        })
+        .ctx(&ctx)
+    }
+}
+
+#[cfg(test)]
```

Comment:
So far my convention has been to move these test into the `serde_arrow/src/test_with_arrow` tree.  As written in the top-level comment, I will add support to simplify getting the data of the arrays so it will be easier to check the contents.

Reply:
sounds good, will move the tests once the helper methods are added